Highlight logic incorrect for checking author set text color

jesxilin opened the pull request:
#### ec22e075245088e425e8c9dfb16efd356e3e2000
<pre>
Highlight logic incorrect for checking author set text color
<a href="https://bugs.webkit.org/show_bug.cgi?id=258943">https://bugs.webkit.org/show_bug.cgi?id=258943</a>
rdar://111706100

Reviewed by Megan Gardner and Aditya Keerthi.

Before logic to determine whether to take a text color of a
StyledMarkedText was to compare to canvasText color.
This causes aproblem if author set color to canvasText or if the color
black was considered as canvasText.
Now, we check if a flag was set indicating the author explicitly
specified a text color.
Updated TestExpectations due to passing test.
Part of priority in spec: <a href="https://www.w3.org/TR/css-highlight-api-1/#priorities">https://www.w3.org/TR/css-highlight-api-1/#priorities</a>

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
(WebCore::coalesceAdjacentWithSameRanges):
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextSelectionPaintStyle):
* Source/WebCore/rendering/TextPaintStyle.h:

Canonical link: <a href="https://commits.webkit.org/266117@main">https://commits.webkit.org/266117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a7a421aef8bc73c5b12c741d13ef2ae6e8257e5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13260 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14674 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13004 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15764 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13282 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15041 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13102 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13804 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10942 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15125 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11092 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11695 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12167 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11862 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15069 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/12332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11594 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/3167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/15913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1469 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/12173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes reviewed (approved):
Looks good! But you'll need someone else with more powers to properly approve

nt1m commented on Source/WebCore/rendering/TextPaintStyle.h:
```diff
@@ -48,6 +48,7 @@ struct TextPaintStyle {
     Color strokeColor;
     Color emphasisMarkColor;
     float strokeWidth { 0 };
+    bool hasExplicitlySetFillColor { false };
```
I would rename this to:
```suggestion
    bool hasExplicitlySetColor { false };
```

because fillColor refers to colors set by `-webkit-text-fill-color` as well (whereas this flag doesn't, it just refers to colors explicitly set by the `color` property).

jesxilin replied:
Yeah, it's a bit confusing but the text color corresponds to the `fillColor` in the `TextPaintStyle` struct, which is why I left it as `fillColor` instead of just `color`.

nt1m replied:
It corresponds to `fillColor` in the `TextPaintStyle`, because I think that color takes in account both `-webkit-text-fill-color` and `color`, while the explicitlySet flag only covers `color`.

nt1m reviewed (commented):
It would be good to know in the commit message which part of the spec this is implementing.

pxlcoder commented on Source/WebCore/rendering/TextPaintStyle.cpp:
```diff
@@ -148,6 +148,9 @@ TextPaintStyle computeTextSelectionPaintStyle(const TextPaintStyle& textPaintSty
     if (emphasisMarkForeground.isValid() && emphasisMarkForeground != selectionPaintStyle.emphasisMarkColor)
         selectionPaintStyle.emphasisMarkColor = emphasisMarkForeground;
 
+    if (auto renderStyle = renderer.parent()->selectionPseudoStyle())
```
Why is this using `renderer.parent()`, and not `renderer`?

jesxilin replied:
`MarkedText::Type::Highlight` takes styles based on `if (auto renderStyle = renderer.parent()->getUncachedPseudoStyle({ PseudoId::Highlight, markedText.highlightName }, &renderer.style()))` so I assumed that it would make sense for selection to follow the same procedure. I'll look into why though.

pxlcoder commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -54,6 +54,7 @@ static StyledMarkedText resolveStyleForMarkedText(const MarkedText& markedText,
             style.backgroundColor = renderStyle->colorResolvingCurrentColor(renderStyle->backgroundColor());
             style.textStyles.fillColor = renderStyle->computedStrokeColor();
             style.textStyles.strokeColor = renderStyle->computedStrokeColor();
```
Not new to this patch, but this looks incorrect? Something to look into later.

```
style.textStyles.fillColor = renderStyle->computedStrokeColor();
```

jesxilin committed:
Highlight logic incorrect for checking author set text color
https://bugs.webkit.org/show_bug.cgi?id=258943
rdar://111706100

Reviewed by Megan Gardner and Aditya Keerthi.

Before logic to determine whether to take a text color of a
StyledMarkedText was to compare to canvasText color.
This causes aproblem if author set color to canvasText or if the color
black was considered as canvasText.
Now, we check if a flag was set indicating the author explicitly
specified a text color.
Updated TestExpectations due to passing test.
Part of priority in spec: https://www.w3.org/TR/css-highlight-api-1/#priorities

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::resolveStyleForMarkedText):
(WebCore::coalesceAdjacentWithSameRanges):
* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextSelectionPaintStyle):
* Source/WebCore/rendering/TextPaintStyle.h:

Canonical link: https://commits.webkit.org/266117@main
